// nexus-prisma/dist-cjs/generator/gentime/settingsSingleton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Gentime = void 0;
const tslib_1 = require("tslib");
const Setset = (0, tslib_1.__importStar)(require("setset"));
var Gentime;
(function (Gentime) {
    Gentime.settings = Setset.create({
        fields: {
            output: {
                shorthand: (directory) => ({ directory }),
                initial: () => ({
                    directory: 'default',
                    name: 'index',
                }),
                fields: {
                    directory: {
                        initial: () => 'default',
                    },
                    name: {
                        initial: () => 'index',
                    },
                },
            },
            projectIdIntToGraphQL: {
                initial: () => 'Int',
            },
            jsdocPropagationDefault: {
                initial: () => 'guide',
            },
            docPropagation: {
                shorthand: (enabled) => ({
                    GraphQLDocs: enabled,
                    JSDoc: enabled,
                }),
                fields: {
                    GraphQLDocs: {
                        initial: () => true,
                    },
                    JSDoc: {
                        initial: () => true,
                    },
                },
            },
            prismaClientImportId: {
                initial: () => `@prisma/client`,
            },
        },
    });
    /**
     * Adjust Nexus Prisma's [gentime settings](https://pris.ly/nexus-prisma/docs/settings/gentime).
     *
     * @example
     *
     *   // prisma/nexus-prisma.ts
     *
     *   import { settings } from 'nexus-prisma/generator'
     *
     *   settings({
     *     projectIdIntToGraphQL: 'ID',
     *   })
     *
     * @remarks This is _different_ than Nexus Prisma's [_runtime_
     *          settings](https://pris.ly/nexus-prisma/docs/settings/runtime).
     */
    function changeSettings(input) {
        Gentime.settings.change(input);
    }
    Gentime.changeSettings = changeSettings;
})(Gentime = exports.Gentime || (exports.Gentime = {}));
