// floggy/build/cjs/level.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parser = exports.LEVELS_BY_NUM = exports.LEVELS = void 0;
exports.LEVELS = {
    fatal: {
        label: `fatal`,
        number: 6
    },
    error: {
        label: `error`,
        number: 5
    },
    warn: {
        label: `warn`,
        number: 4
    },
    info: {
        label: `info`,
        number: 3
    },
    debug: {
        label: `debug`,
        number: 2
    },
    trace: {
        label: `trace`,
        number: 1
    }
};
exports.LEVELS_BY_NUM = Object.values(exports.LEVELS).reduce((lookup, entry) => Object.assign(lookup, { [entry.number]: entry }), {});
/**
 * Parser for log level. The given value is treated case insensitive.
 */
exports.parser = {
    info: {
        typeName: `LogLevel`,
        valid: Object.entries(exports.LEVELS)
            .map(([label]) => label)
            .join(`, `)
    },
    run: (raw) => {
        // eslint-disable-next-line
        return exports.LEVELS[raw.toLowerCase()]?.label ?? null;
    }
};
